src/context/node.js

```diff
@@ -0,0 +1,354 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getMode} from '../mode';
+import {pushIfNotExist, removeItem} from '../utils/array';
+import {startsWith} from '../string';
+import {throttleTail} from './scheduler';
+
+// Properties set on the DOM nodes to track the context state.
+const NODE_PROP = '__AMP_NODE';
+const ASSIGNED_SLOT_PROP = '__AMP_ASSIGNED_SLOT';
+const AMP_PREFIX = 'AMP-';
+
+// Relevant node types.
+// See https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType.
+const ELEMENT_NODE = 1;
+const DOCUMENT_NODE = 9;
+// Includes shadow root, template, etc.
+const FRAGMENT_NODE = 11;
+
+/**
+ * The context node is a sparse tree over the DOM tree. Any node that needs
+ * to manage and compute state can be attached to the context node tree. The
+ * tree (mostly) automatically self-manages: the new nodes and DOM mutations
+ * are auto-discovered or prompted.
+ *
+ * @package
+ */
+export class ContextNode {
+  /**
+   * Returns the existing context node or creates a new one.
+   *
+   * @param {!Node} node
+   * @return {!ContextNode}
+   */
+  static get(node) {
```

Comment:
Yes. And quite often. Typically it signifies a boundary of some sort. One good example are `<slot>` elements would often get a context node since they are on the boundary between shadow DOM and light DOM. Accordion sections would get their context nodes. Fixed elements are likely candidates. Etc. There's absolutely no requirement for context nodes to only be root/AMP. `Closest` simply can short-circuit search when it knows that an element will have context node sooner or later.

Reply:
thank you for the explanation!

Reply:
Sort of. There's not even a way to create a node directly. It's created automatically when a property or a component is added to it. It's pretty similar how there's a DOM tree and a style tree, but you only ever see a DOM tree.